FreeRTOS/FreeRTOS-Plus-TCP
# IPv4/Single: Add a SocketID to a socket

Description
-----------
In [issue #540](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/issues/543), [Jabob Carver](https://github.com/karver8) is asking for the possibility to assign a unique ID to each socket.

We found that a good idea and this PR adds it, depending on the new macro `ipconfigUSE_SetSocketID`.

In many applications, sockets are embedded in an object, a class. The "SetSocketID" can be a pointer to the instance of that class.
These will be the new functions:
~~~c
BaseType_t xSocketSetSocketID( const Socket_t xSocket, void * pvSocketID );
void * pvSocketGetSocketID( const ConstSocket_t xSocket );
~~~

Test Steps
-----------
~~~c
/* Read the current value of the SocketID. */
void * id1 = pvSocketGetSocketID( xSocket );

/* Set value of the SocketID. */
int rc = xSocketSetSocketID( xSocket, ( void * ) 0x12345678 );

/* Read the current value of the SocketID. */
void * id2 = pvSocketGetSocketID( xSocket );

FreeRTOS_printf( ( "SocketSetSocketID %p -> %p rc = %d", id1, id2, ( int ) rc ) );
~~~


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
